### ISSUE TITLE
False positive: C402 Unnecessary generator - rewrite as a dict comprehension.

### ISSUE THREAD

**Issue by jdufresne** [adamchainz/flake8-comprehensions#48]
test.py

```python
s = 'a=1,b=2,c=3'
d = dict(pair.split('=') for pair in s.split(','))
print(d)
```

```
(venv) $ flake8 test.py 
test.py:2:5: C402 Unnecessary generator - rewrite as a dict comprehension.
```

Expected no warnings or errors.

This can't be conveniently rewritten as dict comprehension becuase `pair.split('=')` is an expression, not a simple value. In some cases the expression may be expensive and unwise to execute and index twice.

Discovered by running against the Django code base:

https://github.com/django/django/blob/f3427ccb25806bb062becad0d6a0904930b36635/django/contrib/auth/hashers.py#L381

**Comment by adamchainz** [comment]
Aha, I'm sure we can just tighten C402 to look for a two-tuple as the part of the generator before recommending a dict comprehension

**Comment by jdufresne** [comment]
There was another line discovered in the Django code by this package that wasn't exactly a 2-tuple, but was easy to convert to dict comprehension. So I hope a fix for this doesn't create a new false negative.

From [Django](https://github.com/django/django/blob/master/django/db/migrations/autodetector.py#L1068-L1071):

```python
old_options = dict(
    option for option in old_model_state.options.items()
    if option[0] in AlterModelOptions.ALTER_OPTION_KEYS
)
```

Can be rewritten as dict comprehension:

```python
old_options = {
    key: value for key, value in old_model_state.options.items()
    if key in AlterModelOptions.ALTER_OPTION_KEYS
}
```

**Comment by adamchainz** [comment]
As I wrote in review of #53 and have followed up on in #54, we can't in general convert non-2-tuple generators to dict comprehensions. The example you found there is easier to leave as a false negative, which I think is preferable to leaving the plugin open to more false positives. The exact example could *theoretically* be spotted with a rule to match `.items()` and then spotting the items themselves are being yielded from the generator untouched, but it's quite complicated.
